---
title: "scRNAseq_trial"
author: "Ileana"
date: '2020-03-24'
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Hello World!

I've set-up my first R Notebook and published on Github pages <https://ileana-co.github.io/scRNAseq_trialanalysis/>. Here, I will document my progress on setting up the R Notebook and my attempt to process scRNAseq data. This is a work in progress and meant for organizing what I've learned and could maybe useful for others in the future. Hopefully, it also helps keep me sane and helps keep my writing and communication skills up! #Ilovedocumentation 

Once I figure out how to make a navbar and individual pages, I will organize this page better...but for now I'll leave it all on one page! 

What this page will not cover:

*   How to set-up git and SSH keys to link local computer to repository on git. You can check out all of that [here](https://help.github.com/en/github/getting-started-with-github)
*   How to set-up remote access to BioZone servers. I had to set this up since my lovely MacBook Pro only has 8gb of RAM, but the dataset we are using is already 3gb in size.
*   Common commands for using Power Shell on your terminal and using Markdown 
*   Basics of using R - there are good tutorials out there. I learned this through a course at UofT (EES1137) therefore I am not focusing on documenting my progress of learning this here.

## Setting up R Markdown: trial

After following instructions on <https://resources.github.com/whitepapers/github-and-rstudio/>, here is the start of my Rnotebook set-up. Just trying things out now!

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see [this](http://rmarkdown.rstudio.com).

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. Once you click the **Knit** button to modify the HTML output, you must commit local changes to Git by doing:

```
git add .

git commit -m "comment"

git push -u origin
```

Then go to github to merge your pull request, refresh the webpage, then you are done! Note that you 

You can embed an R code chunk by clicking **Cmd+Option+I (for OS X) or Ctrl +Alt+I**. Rmarkdown/Rnotebook will automatically create a chunk bordered by the delimeters "```{r} code chunk '''"

You can also run each code chunk, which makes Rmarkdown/Rnotebook really convenient for doing your analysis and documenting at the same time. I am personally used to writing scripts in Shell but because of the way R works (processes all code at once into RAM), it's computationally easier to do things this way.

For example:

```{r cars}
summary(cars) # The first trial was to just run the code on Rmarkdown and publish
a <- 1
a # The second was to make edits to the file and publish
```

```{r pressure, echo=FALSE}
plot(pressure) # The third was to generate a plot 
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

## scRNAseq data: Who, What, Where, When, Why?

###**Who?** 
Myself, following a bunch of tutorials from:

*   [Brendan Innes of the Bader Lab at UofT](https://github.com/BaderLab/scRNAseqWorkflow)
*   [Satija Lab (Seurat Package geniuses)](https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html)
*   [The Broad Institute](https://broadinstitute.github.io/2019_scWorkshop/data-wrangling-scrnaseq.html#filtering-low-quality-cells) (*thanks Simon for the link*)
*   [CSC IT Center for Science in Finland and NBISweden by Åsa Björklund and others](https://nbisweden.github.io/excelerate-scRNAseq/session-qc/Quality_control.html)


###**What?** 
To learn, I am using the CRA001160 dataset from the [GSA website](https://bigd.big.ac.cn/search?dbId=gsa&q=CRA001160) (*thanks for the tip JC*) which contains human single-cell RNA sequencing data using an Illumina HiSeq X Ten of:

*   24 human PDAC tumours (male and female)
*   11 normal pancreas (male and female)

###**Where?** 
My home in Toronto, Canada! COVID-19 has taken over the world, labs, businesses and non-essential services have shut down!! Also, "where" I do my analysis is through Rstudio. I started out doing things in R using the terminal since there is more computing power, but in order to document things I'm using Rstudio to access the RNotebook. 

To deal with the large dataset, I have had to use workarounds (for example, the [ff library in R](https://rstudio-pubs-static.s3.amazonaws.com/72295_692737b667614d369bd87cb0f51c9a4b.html)) and  also set-up access to a LINUX machine with large computing power (512gb) operating on CentOS7. These tools will of course depend on the size of the dataset you have. Generally, R is really bad/slow at processing medium (1-2gb) and large (3gb) datasets-- so keep this in mind!

###**When?** 
When I have time and am not distracted :)

###**Why?** 
Analyzing scRNAseq is valuable skill to have especially since I am planning on performing this once the lab gets back up and running again (*who knows how long that is going to take though :( *)) This is especially valuable for my research since mining human datasets of PDAC can help me generate relevant hypotheses for my project... which currently does not have any!!

## Getting Started

### 1. Downloading R and your data!

Assuming you already have installed the [latest version of R](https://www.r-project.org) or Rstudio, downloaded your dataset, you are ready to go! I am currently running R 3.6.3. 

However, I struggled with also just figuring out what these dataset file types are. Typically, you'd want to be able to download the aligned count matrix (if using 10X Genomics, then this is directly from the Cell Ranger pipeline). I believe that Cell Ranger spits out a file containing matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv files. 

*   Matrix.mtx is your count-matrix in sparse matrix format. A sparse matrix essentially is a way to store huge matrices whose contents are mostly zeroes- this is true for count matrices generated by scRNAseq. Sparse matrices basically only store non-zeroes and is really helpful for saving time and computing power. 
*   Genes.tsv and barcodes.tsv basically contain the list of genes and barcodes you will need to make sense of your data

In my case, I was able to download the `count-matrix.txt` file from the GSA database. This was a huge text file of 3-4gb, with different genes/features as row names (already in Gene ID format) and barcodes for each cell as column names. I will discuss below how I dealt with this large set of data.

#### Option 1: Use the ff library 
[ff](https://rstudio-pubs-static.s3.amazonaws.com/72295_692737b667614d369bd87cb0f51c9a4b.html) is the library that will help you deal with large datasets when you have limited RAM on your machine. Basically, ff uses a pointer to a C++ data structure. The location of this pointer is then stored on your RAM and you can access the data set without reading the whole thing into R.

```{r ff, echo = FALSE}

library(ff) # Load the library

## These lines of codes are non-executable here because it takes too much computing power. I have executed these elsewhere

# Read the .txt file and create an ff object
# countMatrix.ff <- read.table(file = "count-matrix.txt") 
# Note that I did not put a path for the file since I am conducting all of this in my current directory. If you are conducting it in another directory you would put (file = "~/ileana/...directory"), for example.

## If your matrix is quite big and you don't really need to analyze all the cells then you can subset your matrix. This makes life much easier. 
# countMatrix <- countMatrix.ff[,1:5000] 

## Take all rows (genes) of only 5000 cells (columns) for your analysis, for example OR use grepl to select certain columns of cells if they are named. 
#countMatrix <- countMatrix.ff[,grepl("T10|T11|T12|T13|T14", colnames = TRUE)] 

## grepl() only takes columns with the labels T10, T11, T12, T13 or T14, effectively subsetting the data. Not sure if this is the best strategy as if I say grepl("T1") then it takes any column with T1 so even T11,T12,T13... but I will stick with this for now

# countMatrix[1:5,1:5] # Display countMatrix to see how your countmatrix looks like
# dim(countMatrix) # Display matrix dimensions
# object.size(countMatrix) # size in bytes


```

This might take a while... It takes my MacBook almost 20 mins and uses 16gb of RAM (using a LOT of swap memory to do so since I only have 8gb of RAM). I also had to basically close every single program I had open! 


### 2. Install the necessary packages
```{r}

# install.packages(c("dplyr","multest","patchwork","Seurat","BiocManager","devtools","Matrix", "ff"))

# BiocManager::install(c("scran","AnnotationDbi","org.Mm.eg.db","org.Hs.eg.db")) # org.Mm for mouse and org.Hs for human

# You might run into some issues installing packages if you are not on the latest version of R. It is best to update R before starting anything!

```

### 3. Load the necessary packages 
```{r}

#library(Seurat)
#library(dplyr)
#library(patchwork)
#library(Matrix)

```

Some info on Seurat and Matrix:

*   Seurat is a very important library developed by the Satija lab and is widely used in all RNAseq analysis. What's important about Seurat is:
  *   It has functions to directly process your 10X data via `Read 10X` (input is the file from Cell Ranger)
  *   You can transform your count matrix into a `Seurat` object which converts the count matrix into a sparse matrix (excludes zero elements) and also serves a container for both data and analysis (clustering results etc). For example, the count matrix is stored in countMatrix[["RNA"]]@counts. See [this page](https://github.com/satijalab/seurat/wiki) for more info on `Seurat` objects
*   Matrix has built in functions to help process the sparse matrix which Seurat converts your data into

